// eclint/dist/rules/insert_final_newline.js
"use strict";
var _ = require('lodash');
var doc = require("../doc");
var EditorConfigError = require("../editor-config-error");
var newlines = {
    lf: '\n',
    crlf: '\r\n',
    cr: '\r'
};
function resolve(settings) {
    if (_.isBoolean(settings.insert_final_newline)) {
        return settings.insert_final_newline;
    }
    return void (0);
}
function check(settings, document) {
    var configSetting = resolve(settings);
    var inferredSetting = infer(document);
    if (configSetting == null || inferredSetting === configSetting) {
        return [];
    }
    var message;
    if (configSetting) {
        message = 'expected final newline';
    }
    else {
        message = 'unexpected final newline';
    }
    var error = new EditorConfigError([message]);
    error.lineNumber = document.lines.length;
    var lastLine = document.lines[document.lines.length - 1];
    error.columnNumber = lastLine.text.length + lastLine.ending.length;
    error.rule = 'insert_final_newline';
    error.source = lastLine.text + lastLine.ending;
    return [error];
}
function fix(settings, document) {
    var lastLine;
    var ending;
    var configSetting = resolve(settings);
    if (configSetting) {
        var endOfLineSetting = settings.end_of_line || 'lf';
        ending = newlines[endOfLineSetting];
    }
    else {
        ending = '';
    }
    while ((lastLine = document.lines[document.lines.length - 1]) && !lastLine.text) {
        document.lines.pop();
    }
    if (lastLine) {
        lastLine.ending = ending;
    }
    else {
        lastLine = new doc.Line({
            number: 1,
            text: '',
            ending: ending,
            offset: 0
        });
        document.lines.push(lastLine);
    }
    return document;
}
function infer(document) {
    var lastLine = document.lines[document.lines.length - 1];
    if (lastLine && lastLine.ending) {
        return true;
    }
    return false;
}
var InsertFinalNewlineRule = {
    type: 'DocumentRule',
    resolve: resolve,
    check: check,
    fix: fix,
    infer: infer
};
module.exports = InsertFinalNewlineRule;
